scrapypuppeteer/middleware.py

```diff
@@ -164,3 +166,120 @@ def close_used_contexts(self, spider):
+        if recaptcha_solving and not submit_selectors:
+            raise ValueError('No submit selectors provided but automatic solving is enabled')
+        return cls(recaptcha_solving, submit_selectors)
+
+    @staticmethod
+    def process_request(request, spider):
+        # We don't modify any request, we only work with responses
+        return None
+
+    def process_response(self,
+                         request, response,
+                         spider):
+        # TODO: What if we did not provide 2captcha token?
+        if not isinstance(response, PuppeteerResponse):  # We only work with PuppeteerResponses
+            return response
+
+        if response.puppeteer_request.close_page:  # No need in solving captcha right before closing the page
+            return response
+
+        if isinstance(response.puppeteer_request.action, Click) and \
+                response.puppeteer_request.action.selector in self.submit_selectors.values():  # Submitted captcha
+            return self._gen_response(response)
+
+        if isinstance(response.puppeteer_request.action, RecaptchaSolver):
+            # RECaptchaSolver was called by recaptcha middleware
+            return self._submit_recaptcha(request, response, spider)
+
+        # Any puppeteer response besides RecaptchaSolver's PuppeteerResponse
+        return self._solve_recaptcha(request, response)
+
+    def _solve_recaptcha(self, request, response):
+        recaptcha_solver = RecaptchaSolver(solve_recaptcha=self.recaptcha_solving)
+        self._page_responses[response.page_id] = response  # Saving main response to return it later
+        return response.follow(recaptcha_solver,
+                               callback=request.callback,
+                               errback=request.errback,
+                               close_page=False)
+
+    def _submit_recaptcha(self, request, response, spider):
+        response_data = response.data
+        if not response.puppeteer_request.action.solve_recaptcha:
+            spider.log(message=f"Found {len(response_data['recaptcha_data']['captchas'])} captcha "
+                               f"but did not solve due to argument",
+                       level=logging.INFO)
+            return self._gen_response(response)
+        # We need to click "submit button"
+        for domain, submit_selector in self.submit_selectors.items():
+            if domain in response.url:
+                if not submit_selector:
+                    return self._gen_response(response)
+                submit_click = Click(submit_selector)
+                return response.follow(action=submit_click,
+                                       callback=request.callback,
+                                       errback=request.errback,
+                                       close_page=response.puppeteer_request.close_page)
+        raise IgnoreRequest("No submit selector found to click on the page")
```

Comment:
This will be raised on all non-captcha responses unless their urls are whitelisted in `submit_selectors`